File: ruby-pattern/FactoryMethod/factory_method.rb

class Product

  def initialize
  end

end

class ProductFactory

  def initialize
  end

  def create_product
    pre_create
    product = create
    post_create
    return product
  end

end

class ConcreteProductA < Product
end

class ConcreteProductAFactory < ProductFactory

  def pre_create
    print "pre A create\n"
  end

  def create
    print "A create\n"
    return ConcreteProductA.new
  end

  def post_create
    print "post A create\n"
  end
end

factory = ConcreteProductAFactory.new

factory.create_product

File: swift-pattern/design-pattern.playground/Sources/FactoryMethod.swift
public class Product{
    var id:String = ""
    init(id:String) {
        self.id = id
    }
    public func productId()->String{
        return id
    }
}

public class BaseFactory{

    public init() {}
    public func preCreate(){}
    public func create()->Product{return Product.init(id: "")}
    public func postCreate(){}

    public func createProduct()->Product{
        preCreate()
        let product:Product = create()
        postCreate()
        return product
    }
}

public class ConcreteFactoryA:BaseFactory{
    public override init() {}
    public override func preCreate() {
        print("something before create in A")
    }
    public override func create() -> Product {
        print("something creating in A")
        return Product.init(id: "from A Factory")
    }
    public override func postCreate() {
        print("something after create in A")
    }
}

public class ConcreteFactoryB:BaseFactory{
    public override init() {}
    public override func preCreate() {
        print("something before create in B")
    }
    public override func create() -> Product {
        print("something creating in B")
        return Product.init(id: "from B Factory")
    }
    public override func postCreate() {
        print("something after create in B")
    }
}




